### PR TITLE
move CapAdd and PublishPorts config settings to #start method on container instead of #create

### PR DESCRIPTION
**Pull request by talevy**

### CODE REVIEW

**Comment by talevy** [comment]
not sure how it worked for you, but this was required for me

File: lib/longshoreman/container.rb [hunk 1]
```diff
@@ -76,7 +74,7 @@ def rport(exposed_port, protocol="tcp")
   end
 
   def start
-    @raw.start
+    @raw.start({ 'PublishAllPorts' => true, 'CapAdd' => 'NET_ADMIN' })
```
Comment (untergeek) [inline lib/longshoreman/container.rb:77]: Why?  I know it works with PublishAllPorts in main_args.  Does the CapAdd not work there?

**Comment by untergeek** [comment]
Probably because you're creating an image from a Dockerfile, and a container from the result, though I can't explain how/why that would be different. 

I've never needed it in the start, but we can do that, if it works both ways.

File: lib/longshoreman/container.rb (hunk 1 above)
Comment (talevy) [inline lib/longshoreman/container.rb:77]: for my zookeeper container/image I was receiving this as the network settings before this fix

```
"Ports"=>{"2181/tcp"=>nil, "2888/tcp"=>nil, "3888/tcp"=>nil}
```

after fix

```
{"2181/tcp"=>[{"HostIp"=>"0.0.0.0", "HostPort"=>"49156"}], "2888/tcp"=>[{"HostIp"=>"0.0.0.0", "HostPort"=>"49157"}], "3888/tcp"=>[{"HostIp"=>"0.0.0.0", "HostPort"=>"49158"}]}
```

Comment (talevy) [inline lib/longshoreman/container.rb:77]: and for docker-doctor, this is where I had set these settings, and they worked.

**Comment by untergeek** [comment]
If that tests out, version bump and I'll merge.

**Comment by talevy** [comment]
no, in this case I am using an existing image that I previously pulled into my local docker.

**Comment by talevy** [comment]
example:

```
require 'longshoreman'

ZK_EXPOSED_PORT = 2181
KAFKA_EXPOSED_PORT = 9092

extra = {"ExposedPorts" => { "#{KAFKA_EXPOSED_PORT}/tcp" => {} } }
zk = Longshoreman::Container.new(image='jplock/zookeeper:latest', name='zookeeper')
k1 = Longshoreman::Container.new(image='wurstmeister/kafka:0.8.2.1', name='k1', extra_args=extra)
k2 = Longshoreman::Container.new(image='wurstmeister/kafka:0.8.2.1', name='k2', extra_args=extra)
k3 = Longshoreman::Container.new(image='wurstmeister/kafka:0.8.2.1', name='k3', extra_args=extra)
zk_port = zk.rport(ZK_EXPOSED_PORT)
k1_port = k1.rport(KAFKA_EXPOSED_PORT)
k2_port = k2.rport(KAFKA_EXPOSED_PORT)
k3_port = k3.rport(KAFKA_EXPOSED_PORT)

[zk, k1, k2, k3].map { |x| x.cleanup }
```

**Comment by untergeek** [comment]
Hmmm.  That's what I think is going on.  The "ExposedPorts" is being assigned _outside_ the image, rather than as a part of the image, and that's why it needs to be in the start statement.

No worries, though. If it works your way for all situations we'll go with that.

**Comment by talevy** [comment]
no :)

that only applies to kafka's container. I initially ran into this with the zookeeper container, which exposes the port within the Dockerfile.

**Comment by untergeek** [comment]
Weird.  Docker's inconsistent, if that's the case.  The reason I put it into the args is that's what the [example on the Docker Remote API page says to do](https://docs.docker.com/reference/api/docker_remote_api_v1.18/#create-a-container).  All of these arguments are in the create block.

**Comment by talevy** [comment]
no clue, I am using docker 1.5. let me upgrade and see if that changes anything

**Comment by untergeek** [comment]
Ah, I'm on 1.6.

**Comment by untergeek** [comment]
How will this resolve?

**Comment by talevy** [comment]
hmm. I can't reproduce this anymore on 1.6 in normal docker....

I'll go ahead and close this for now

**Pull request closed without merging**